### ISSUE TITLE
Bug: SEGV in smart-cast block

### ISSUE THREAD

**Issue by sekiguchi-nagisa** [sekiguchi-nagisa/ydsh#687]
## Impact Target
* v0.29.0
* v0.30.0-unstable

## Step
```
var a = "34".toInt()
if $a {
  $MODULE._func('$a = $none')()
  $a++
  echo $a
}
```

## Problem
in smart-cast block, `a` should always indicate Int value, but after call `Module#_func`, `a` indicate invalid value

**Comment by sekiguchi-nagisa** [comment]
should restrict smart-cast condition
* if read only local/global variables, enable smart cast
* if local variables that are not captured, enable smart cast
* smart-cast of global variables should not be allowed

**Comment by sekiguchi-nagisa** [comment]
fix in commit ec647da
